magma/syntax/util.py

```diff
@@ -1,18 +1,37 @@
+from typing import Optional, Mapping, Sequence, Any
+
 from ..t import In, Out
 from ..circuit import Circuit
 from ..wire import wire
 
 
-def build_io_args(annotations):
+def build_io_args(
+        annotations: Mapping[str, Any],
+        output_port_names: Optional[Sequence[str]] = None,
+        ) -> Mapping[str, Any]:
     io_args = {}
 
     for param, annotation in annotations.items():
         if param == "return":
             if isinstance(annotation, tuple):
-                for i, elem in enumerate(annotation):
-                    io_args[f"O{i}"] = Out(elem)
+                if output_port_names is None:
+                    for i, elem in enumerate(annotation):
+                        io_args[f"O{i}"] = Out(elem)
+                elif len(output_port_names) == len(annotation):
+                    for name, elem in zip(output_port_names, annotation):
+                        if not isinstance(name, str):
+                            raise TypeError(f'port names must be strings recieved {name}::{type(name)}')
+                        io_args[name] = Out(elem)
+                else:
+                    raise ValueError(f'recieved {len(output_port_names)} names expected {len(annotation)}')
             else:
-                io_args["O"] = Out(annotation)
+                if output_port_names is None:
+                    io_args["O"] = Out(annotation)
+                elif len(output_port_names) == 1:
+                    io_args[output_port_names[0]] = Out(annotation)
+                else:
+                    raise ValueError(f'recieved {len(output_port_names)} names expected 1')
+
```

Comment:
This might be simpler if we filled output_port_names with the default name(s) (i.e. "O"` or "O{i}") if the sequence isn't provided, then the logic would be the same for both cases? Probably not that much of an improvement if any though.

Reply:
Yeah I actually thought about doing that but there are multiple call sites and I was feeling lazy.  I can move the building of the out_port_names to the top though.